File: js/render.mjs
// js/render.mjs — the Planes canonical renderer, ported from render.py.
//
// render(parse(src)) must parse to an equal AST for every program in the repo —
// not byte-identical to the original source, but canonical: the same formatting
// decision everywhere, every call rendered `name of (args)`, every compound
// sub-expression parenthesised. Source position is not preserved, so round-trip
// equality is checked with astEqual, which ignores `line`.
//
// A.4: every AST node kind has a real case, no safe fallback — an unhandled kind
// raises, naming it (the two node kinds that were silently unrenderable in
// render.py were caught exactly because there was no fallback to hide them).
// Checked against render.py's output byte for byte across the corpus, with a
// per-node-kind round-trip and cross-implementation round-trips both directions
// (test_js_render.py). render.py is the specification.
//
// Browser-safe (A.7): imports rules.mjs (itself browser-safe) for the marker
// path, and nothing Node-only.

import { escapeStringLiteral } from "./planes_text.mjs";
import { isNode, isTup } from "./nodes.mjs";
import { PlanesNumber } from "./planes_num.mjs";
import { check } from "./rules.mjs";

const INDENT = "  ";

// Sub-expressions that need parens wherever they are not the outermost
// expression of their statement.
const COMPOUND = new Set(["BinOp", "Not", "IsNothing"]);

// Expression node kinds that may stand alone as a statement. Listed explicitly
// so renderStmt raises on a node that is neither a known statement nor a
// renderable expression, rather than routing an unknown node through renderExpr.
const EXPR_STMT = new Set([
  "Num", "Str", "Bool", "Nothing", "Var", "ListLit", "RecordLit",
  "RecordUpdate", "ListPlus", "BinOp", "Not", "IsNothing", "Field", "Call",
  "Round", "ForEach", "OrFail",
]);

// A RecordLit/RecordUpdate/Use/Note field Tup — its items.
function parts(t) {
  return isTup(t) ? t.items : t;
}

// ================================================================ expressions

function renderOperand(node) {
  const text = renderExpr(node);
  return COMPOUND.has(node.__node) ? `(${text})` : text;
}

export function renderExpr(node) {
  switch (node.__node) {
    case "Num":
      return node.value && typeof node.value.text === "function"
        ? node.value.text()
        : String(node.value);
    case "Str":
      return `"${escapeStringLiteral(node.value)}"`;
    case "Bool":
      return node.value ? "true" : "false";
    case "Nothing":
      return "nothing";
    case "Var":
      return node.name;
    case "ListLit":
      return "[" + node.items.map(renderExpr).join(", ") + "]";
    case "ListPlus":
      return `${renderOperand(node.base)} plus ${renderOperand(node.item)}`;
    case "RecordLit": {
      const fields = node.fields
        .map((f) => {
          const [k, v] = parts(f);
          return `${k}: ${renderExpr(v)}`;
        })
        .join(", ");
      return fields ? "{ " + fields + " }" : "{}";
    }
    case "RecordUpdate": {
      const fields = node.fields
        .map((f) => {
          const [k, v] = parts(f);
          return `${k}: ${renderExpr(v)}`;
        })
        .join(", ");
      return `${renderOperand(node.base)} with ${fields}`;
    }
    case "BinOp":
      if (node.op === "first") {
        return `first ${renderOperand(node.left)} of ${renderOperand(node.right)}`;
      }
      return `${renderOperand(node.left)} ${node.op} ${renderOperand(node.right)}`;
    case "Not":
      return `not ${renderOperand(node.expr)}`;
    case "IsNothing":
      return `${renderOperand(node.expr)} is nothing`;
    case "Field":
      return `${renderOperand(node.obj)}.${node.name}`;
    case "Call":
      return renderCall(node);
    case "Round":
      return `round ${renderOperand(node.value)} to ${renderOperand(node.places)} places`;
    case "ForEach":
      return renderForeachExpr(node);
    case "OrFail":
      return renderOrfail(node);
    case "Builtin":
      // Dead: the parser never builds this node. Named failure, not a silent
      // fall-through, if that ever stops being true.
      throw new Error("renderExpr: Builtin is unreachable by design");
    default:
      throw new Error(`renderExpr: unhandled node type ${node.__node}`);
  }
}

function renderCall(node) {
  if (!node.args.length) return node.name;
  const args = node.args.map((a) => `(${renderExpr(a)})`).join(", ");
  return `${node.name} of ${args}`;
}

function renderForeachExpr(node) {
  const where = node.where !== null && node.where !== undefined
    ? ` where ${renderExpr(node.where)}`
    : "";
  const body = renderExpr(node.body[0]);
  return `for each ${node.var} in ${renderExpr(node.source)}${where}: ${body}`;
}

function renderWritetoInline(node) {
  return `write ${renderExpr(node.value)} to ${renderExpr(node.dest)}`;
}

function renderOrfail(node) {
  const inner =
    node.expr.__node === "WriteTo"
      ? renderWritetoInline(node.expr)
      : renderExpr(node.expr);
  return `${inner} or fail as ${node.tag}`;
}

// ================================================================ statements

function renderBecauseSuffix(node) {
  if (node.annotation === null || node.annotation === undefined) return "";
  return ` because "${escapeStringLiteral(node.annotation.text)}"`;
}

function renderAssign(node) {
  const prefix = node.is_let ? "let " : "";
  return `${prefix}${node.name} = ${renderExpr(node.expr)}${renderBecauseSuffix(node)}`;
}

function renderRule(node) {
  const verb = node.assertion === "forbid" ? "may not" : "may";
  let text = `rule [${node.name}] ${node.subject} ${verb} ${node.kind}`;
  if (node.target !== null && node.target !== undefined) {
    text += ` to "${escapeStringLiteral(node.target)}"`;
  }
  if (node.supersedes !== null && node.supersedes !== undefined) {
    text += ` supersedes [${node.supersedes}]`;
    if (node.supersedes_fingerprint !== null && node.supersedes_fingerprint !== undefined) {
      text += ` @${node.supersedes_fingerprint}`;
    }
  }
  return text + renderBecauseSuffix(node);
}

function renderNote(node, indent) {
  const lines = [indent + "note:"];
  for (const e of node.entries) {
    const [kind, value] = parts(e);
    if (kind === "from") {
      lines.push(indent + INDENT + `from "${escapeStringLiteral(value)}"`);
    } else if (kind === "derives-from") {
      lines.push(indent + INDENT + `derives-from [${value}]`);
    }
  }
  return lines.join("\n");
}

function renderUse(node) {
  let text = `use ${node.module}`;
  for (const r of node.renames) {
    const [old, nw] = parts(r);
    text += ` with ${old} as ${nw}`;
  }
  return text;
}

function renderForeign(node) {
  let text = `foreign ${node.name}`;
  if (node.params.length) text += " of " + node.params.join(", ");
  text += ` from "${escapeStringLiteral(node.target)}"`;
  if (node.declared) {
    if (!node.effects.length) {
      text += " doing nothing";
    } else {
      const claims = [];
      for (const eff of node.effects) {
        const [kind, where] = parts(eff);
        if (where === null || where === undefined) {
          claims.push(kind);
        } else if (where.items[0] === "literal") {
          claims.push(`${kind} "${escapeStringLiteral(where.items[1])}"`);
        } else {
          // ("param", name)
          claims.push(`${kind} ${where.items[1]}`);
        }
      }
      text += " doing " + claims.join(", ");
    }
  }
  return text;
}

function renderFuncdef(node, indent, markers) {
  let header = `to ${node.name}`;
  if (node.params.length) header += " of " + node.params.join(", ");
  header += ":";
  const body = renderBlock(node.body, indent + INDENT, markers);
  return indent + header + "\n" + body;
}

function renderIf(node, indent, markers) {
  const lines = [indent + `if ${renderExpr(node.cond)}:`];
  lines.push(renderBlock(node.then, indent + INDENT, markers));
  if (node.els.length) {
    lines.push(indent + "else:");
    lines.push(renderBlock(node.els, indent + INDENT, markers));
  }
  return lines.join("\n");
}

function renderForeachStmt(node, indent, markers) {
  const where = node.where !== null && node.where !== undefined
    ? ` where ${renderExpr(node.where)}`
    : "";
  const header = indent + `for each ${node.var} in ${renderExpr(node.source)}${where}:`;
  const body = renderBlock(node.body, indent + INDENT, markers);
  return header + "\n" + body;
}

function renderWhen(node, indent, markers) {
  const entries = [];
  for (const p of node.pattern) {
    const [fname, inner] = p.items;
    const [kind, arg] = inner.items;
    entries.push(kind === "match" ? `${fname}: ${renderExpr(arg)}` : fname);
  }
  const header = indent + `when ${renderExpr(node.subject)} is { ${entries.join(", ")} }:`;
  const lines = [header, renderBlock(node.body, indent + INDENT, markers)];
  if (node.els.length) {
    lines.push(indent + "else:");
    lines.push(renderBlock(node.els, indent + INDENT, markers));
  }
  return lines.join("\n");
}

export function renderStmt(node, indent, markers) {
  switch (node.__node) {
    case "Use":
      return indent + renderUse(node);
    case "Foreign":
      return indent + renderForeign(node);
    case "Rule":
      return indent + renderRule(node);
    case "Note":
      return renderNote(node, indent);
    case "FuncDef":
      return renderFuncdef(node, indent, markers);
    case "Assign":
      return indent + renderAssign(node);
    case "Give":
      return indent + `give ${renderExpr(node.expr)}`;
    case "Show":
      return indent + `show ${renderExpr(node.expr)}`;
    case "Why":
      return indent + `why ${renderExpr(node.expr)}`;
    case "When":
      return renderWhen(node, indent, markers);
    case "WriteTo":
      return indent + renderWritetoInline(node);
    case "OrFail":
      return indent + renderOrfail(node);
    case "Fail":
      return indent + `fail ${renderExpr(node.message)} as ${node.tag}`;
    case "If":
      return renderIf(node, indent, markers);
    case "ForEach":
      return renderForeachStmt(node, indent, markers);
    default:
      // A bare expression statement. Every expression node kind is dispatched
      // explicitly by renderExpr; a node that is neither a statement above nor
      // a renderable expression raises here, naming the kind (A.4).
      if (EXPR_STMT.has(node.__node)) return indent + renderExpr(node);
      throw new Error(`renderStmt: unhandled node type ${node.__node}`);
  }
}

// ================================================================ the generated marker

// Every source line this node's subtree touches — the AST-native replacement
// for scanning text. Matches render.py's _line_span: recurses into direct node
// fields and one level into list/tuple elements that are themselves nodes (a
// Tup element, being no node, is not descended — a faithful quirk).
function lineSpan(node, seen) {
  seen = seen ?? new Set();
  if (!isNode(node) || seen.has(node)) return new Set();
  seen.add(node);
  const lines = new Set();
  if (node.line) lines.add(node.line);
  for (const f of Object.keys(node)) {
    if (f === "__node") continue;
    const v = node[f];
    if (isNode(v)) {
      for (const x of lineSpan(v, seen)) lines.add(x);
    } else if (Array.isArray(v)) {
      for (const x of v) {
        if (isNode(x)) for (const y of lineSpan(x, seen)) lines.add(y);
      }
    }
  }
  return lines;
}

export function computeMarkers(rules, surface, declaringFile = null) {
  if (!rules || !rules.length) return new Map();
  if (surface === null || surface === undefined) {
    throw new Error(
      "render(): rules given without surface -- markers need a " +
        "computed effect surface\n" +
        "  try: render(prog, rules=found, surface=analyse(src))",
    );
  }
  const results = check(rules, surface, declaringFile);
  const markers = new Map();
  for (const v of results) {
    if (v.effect === null || v.effect === undefined) continue;
    if (!markers.has(v.effect.site)) markers.set(v.effect.site, []);
    markers.get(v.effect.site).push(v.rule.name);
  }
  return markers;
}

function markerLines(stmt, indent, markers) {
  if (!markers.size) return [];
  const span = lineSpan(stmt);
  const hit = [...span].filter((ln) => markers.has(ln)).sort((a, b) => a - b);
  const names = [];
  const seen = new Set();
  for (const ln of hit) {
    for (const name of markers.get(ln)) {
      if (!seen.has(name)) {
        seen.add(name);
        names.push(name);
      }
    }
  }
  return names.map((name) => indent + `~ [${name}] applies here`);
}

function renderBlock(stmts, indent, markers) {
  const out = [];
  for (const s of stmts) {
    out.push(...markerLines(s, indent, markers));
    out.push(renderStmt(s, indent, markers));
  }
  return out.join("\n");
}

// ================================================================ entry points

export function render(prog, rules = null, surface = null) {
  const markers = computeMarkers(rules, surface);
  const body = renderBlock(prog, "", markers);
  return body ? body + "\n" : "";
}

export function stripAnnotations(prog) {
  const stripStmts = (stmts) =>
    stmts.filter((s) => s.__node !== "Note").map(stripOne);
  const stripOne = (s) => {
    if (s.__node === "Assign" && s.annotation !== null && s.annotation !== undefined) {
      return { ...s, annotation: null };
    }
    if (s.__node === "Rule" && s.annotation !== null && s.annotation !== undefined) {
      return { ...s, annotation: null };
    }
    if (s.__node === "If") {
      return { ...s, then: stripStmts(s.then), els: stripStmts(s.els) };
    }
    if (s.__node === "ForEach") return { ...s, body: stripStmts(s.body) };
    if (s.__node === "FuncDef") return { ...s, body: stripStmts(s.body) };
    return s;
  };
  return stripStmts(prog);
}

// Structural equality that ignores `line` — source position, not meaning. The
// canonical renderer legitimately moves line numbers, so round-trip equality
// checks structure, not accidental line agreement.
export function astEqual(a, b) {
  if (isNode(a) || isNode(b)) {
    if (!isNode(a) || !isNode(b) || a.__node !== b.__node) return false;
    for (const f of Object.keys(a)) {
      if (f === "line" || f === "__node") continue;
      if (!(f in b)) return false;
      if (!astEqual(a[f], b[f])) return false;
    }
    for (const f of Object.keys(b)) {
      if (f === "line" || f === "__node") continue;
      if (!(f in a)) return false;
    }
    return true;
  }
  if (isTup(a) || isTup(b)) {
    if (!isTup(a) || !isTup(b)) return false;
    return astEqual(a.items, b.items);
  }
  if (Array.isArray(a) || Array.isArray(b)) {
    if (!Array.isArray(a) || !Array.isArray(b) || a.length !== b.length) return false;
    return a.every((x, i) => astEqual(x, b[i]));
  }
  if (a instanceof PlanesNumber || b instanceof PlanesNumber) {
    return a instanceof PlanesNumber && b instanceof PlanesNumber && a.eq(b);
  }
  return a === b;
}

File: swift/Sources/Planes/Render.swift
// Render.swift — the Planes canonical renderer, ported from render.py.
//
// The Swift counterpart of js/render.mjs, keeping its structure and names.
// render(parse(src)) parses back to an equal AST: canonical, not literal — every
// call rendered `name of (args)`, every compound sub-expression parenthesised.
// With rules and a surface it also emits the generated rule markers
// (`~ [name] applies here`), which is what `render-rules` (shapes_cli.py
// --render) prints; test_swift_rules.py holds that output byte for byte against
// render.py. Every node kind has a real case and no safe fallback: an unhandled
// kind throws, naming it.

/// A node the renderer has no case for.
public struct RenderError: Error, CustomStringConvertible, Sendable {
    public let message: String
    public var description: String { message }
}

private let INDENT = "  "

// Sub-expressions that need parens wherever they are read at less than full
// precedence.
private let COMPOUND: Set<AST.Kind> = [.BinOp, .Not, .IsNothing, .ListPlus, .OrFail, .RecordUpdate, .ForEach]

// Expressions whose own trailing structure swallows a following keyword or `:`
// delimiter.
private let OPEN_TRAILING: Set<AST.Kind> = [.OrFail, .RecordUpdate, .ForEach]

// Expression node kinds that may stand alone as a statement.
private let EXPR_STMT: Set<AST.Kind> = [
    .Num, .Str, .Bool, .Nothing, .Var, .ListLit, .RecordLit, .RecordUpdate, .ListPlus, .BinOp, .Not,
    .IsNothing, .Field, .Call, .Round, .ForEach, .OrFail,
]

private func delimited(_ node: AST.Node) throws -> String {
    OPEN_TRAILING.contains(node.kind) ? "(\(try renderExpr(node)))" : try renderExpr(node)
}

// ================================================================ comma-list elements

// The greedy comma-extensible list that terminates renderExpr(node) at top level:
// "of" (a call's argument list), "with" (a record update's field list), or nil.
private func greedyTail(_ node: AST.Node) -> String? {
    switch node {
    case let c as AST.Call: return c.args.isEmpty ? nil : "of"
    case is AST.RecordUpdate: return "with"
    case let fe as AST.ForEach: return fe.body.first.flatMap(greedyTail)
    case let b as AST.BinOp: return COMPOUND.contains(b.right.kind) ? nil : greedyTail(b.right)
    case let n as AST.Not: return COMPOUND.contains(n.expr.kind) ? nil : greedyTail(n.expr)
    case let l as AST.ListPlus: return COMPOUND.contains(l.item.kind) ? nil : greedyTail(l.item)
    default: return nil
    }
}

// Render `node` as an element of a comma-separated list, parenthesising it when
// its greedy tail would be swallowed by the list's own separator.
private func commaElement(_ node: AST.Node, _ sep: String) throws -> String {
    let text = try renderExpr(node)
    let tail = greedyTail(node)
    let dangerous = tail == "of" || (tail == "with" && sep == "record")
    return dangerous ? "(\(text))" : text
}

// The base of a `X.name` field access.
private func fieldBase(_ node: AST.Node) throws -> String {
    if COMPOUND.contains(node.kind) || greedyTail(node) != nil {
        return "(\(try renderExpr(node)))"
    }
    return try renderExpr(node)
}

// ================================================================ expressions

// A BinOp the parser SYNTHESISED for a unary minus, `-X`: a subtraction from a
// literal zero renders as the unary form whether the zero was synthesised or
// written.
private func isNegation(_ node: AST.BinOp) -> Bool {
    guard sameText(node.op, "-"), let zero = node.left as? AST.Num else { return false }
    return zero.value.isZero()
}

private func renderOperand(_ node: AST.Node) throws -> String {
    let text = try renderExpr(node)
    return COMPOUND.contains(node.kind) ? "(\(text))" : text
}

private func recordFields(_ fields: [(name: String, expr: AST.Node)]) throws -> String {
    try fields.map { "\($0.name): \(try commaElement($0.expr, "record"))" }.joined(separator: ", ")
}

public func renderExpr(_ node: AST.Node) throws -> String {
    switch node {
    case let n as AST.Num: return n.value.text()
    case let s as AST.Str: return "\"\(escapeStringLiteral(s.value))\""
    case let b as AST.Bool: return b.value ? "true" : "false"
    case is AST.Nothing: return "nothing"
    case let v as AST.Var: return v.name
    case let l as AST.ListLit:
        return "[" + (try l.items.map { try commaElement($0, "list") }).joined(separator: ", ") + "]"
    case let l as AST.ListPlus:
        return "\(try renderOperand(l.base)) plus \(try renderOperand(l.item))"
    case let r as AST.RecordLit:
        let fields = try recordFields(r.fieldList)
        return fields.isEmpty ? "{}" : "{ " + fields + " }"
    case let r as AST.RecordUpdate:
        return "\(try renderOperand(r.base)) with \(try recordFields(r.fieldList))"
    case let b as AST.BinOp:
        if sameText(b.op, "first") {
            return "first (\(try renderExpr(b.left))) of (\(try renderExpr(b.right)))"
        }
        if isNegation(b) {
            return "-\(try renderOperand(b.right))"
        }
        return "\(try renderOperand(b.left)) \(b.op) \(try renderOperand(b.right))"
    case let n as AST.Not: return "not \(try renderOperand(n.expr))"
    case let n as AST.IsNothing: return "\(try renderOperand(n.expr)) is nothing"
    case let f as AST.Field: return "\(try fieldBase(f.obj)).\(f.name)"
    case let c as AST.Call: return try renderCall(c)
    case let r as AST.Round: return "round \(try renderOperand(r.value)) to \(try renderOperand(r.places)) places"
    case let fe as AST.ForEach: return try renderForeachExpr(fe)
    case let o as AST.OrFail: return try renderOrfail(o)
    case is AST.Builtin:
        throw RenderError(message: "renderExpr: Builtin is unreachable by design")
    default:
        throw RenderError(message: "renderExpr: unhandled node type \(node.kind.rawValue)")
    }
}

private func renderCall(_ node: AST.Call) throws -> String {
    if node.args.isEmpty { return node.name }
    let args = try node.args.map { "(\(try renderExpr($0)))" }.joined(separator: ", ")
    return "\(node.name) of \(args)"
}

private func renderForeachExpr(_ node: AST.ForEach) throws -> String {
    let whereText = try node.whereClause.map { " where \(try delimited($0))" } ?? ""
    guard let first = node.body.first else {
        throw RenderError(message: "renderExpr: a for-each expression with no body")
    }
    return "for each \(node.variable) in \(try delimited(node.source))\(whereText): \(try renderExpr(first))"
}

private func renderWritetoInline(_ node: AST.WriteTo) throws -> String {
    "write \(try delimited(node.value)) to \(try delimited(node.dest))"
}

private func renderOrfail(_ node: AST.OrFail) throws -> String {
    let inner = try (node.expr as? AST.WriteTo).map(renderWritetoInline) ?? delimited(node.expr)
    return "\(inner) or fail as \(node.tag)"
}

// ================================================================ statements

private func renderBecauseSuffix(_ annotation: AST.Because?) -> String {
    guard let annotation else { return "" }
    return " because \"\(escapeStringLiteral(annotation.text))\""
}

private func renderAssign(_ node: AST.Assign) throws -> String {
    let prefix = node.isLet ? "let " : ""
    return "\(prefix)\(node.name) = \(try renderExpr(node.expr))\(renderBecauseSuffix(node.annotation))"
}

private func renderRule(_ node: AST.Rule) -> String {
    let verb = sameText(node.assertion, "forbid") ? "may not" : "may"
    var text = "rule [\(node.name)] \(node.subject) \(verb) \(node.effectKind)"
    if let target = node.target { text += " to \"\(escapeStringLiteral(target))\"" }
    if let supersedes = node.supersedes {
        text += " supersedes [\(supersedes)]"
        if let fp = node.supersedesFingerprint { text += " @\(fp)" }
    }
    return text + renderBecauseSuffix(node.annotation)
}

private func renderNote(_ node: AST.Note, _ indent: String) -> String {
    var lines = [indent + "note:"]
    for e in node.entries {
        if sameText(e.kind, "from") {
            lines.append(indent + INDENT + "from \"\(escapeStringLiteral(e.value))\"")
        } else if sameText(e.kind, "derives-from") {
            lines.append(indent + INDENT + "derives-from [\(e.value)]")
        }
    }
    return lines.joined(separator: "\n")
}

private func renderUse(_ node: AST.Use) -> String {
    var text = "use \(node.module)"
    for r in node.renames { text += " with \(r.old) as \(r.new)" }
    return text
}

private func renderForeign(_ node: AST.Foreign) -> String {
    var text = "foreign \(node.name)"
    if !node.params.isEmpty { text += " of " + node.params.joined(separator: ", ") }
    text += " from \"\(escapeStringLiteral(node.target))\""
    if node.declared {
        if node.effects.isEmpty {
            text += " doing nothing"
        } else {
            let claims = node.effects.map { eff -> String in
                switch eff.target {
                case nil: return eff.kind
                case let .literal(value)?: return "\(eff.kind) \"\(escapeStringLiteral(value))\""
                case let .param(name)?: return "\(eff.kind) \(name)"
                }
            }
            text += " doing " + claims.joined(separator: ", ")
        }
    }
    return text
}

private func renderFuncdef(_ node: AST.FuncDef, _ indent: String, _ markers: Markers) throws -> String {
    var header = "to \(node.name)"
    if !node.params.isEmpty { header += " of " + node.params.joined(separator: ", ") }
    header += ":"
    return indent + header + "\n" + (try renderBlock(node.body, indent + INDENT, markers))
}

private func renderIf(_ node: AST.If, _ indent: String, _ markers: Markers) throws -> String {
    var lines = [indent + "if \(try delimited(node.cond)):"]
    lines.append(try renderBlock(node.then, indent + INDENT, markers))
    if !node.els.isEmpty {
        lines.append(indent + "else:")
        lines.append(try renderBlock(node.els, indent + INDENT, markers))
    }
    return lines.joined(separator: "\n")
}

private func renderForeachStmt(_ node: AST.ForEach, _ indent: String, _ markers: Markers) throws -> String {
    let whereText = try node.whereClause.map { " where \(try delimited($0))" } ?? ""
    let header = indent + "for each \(node.variable) in \(try delimited(node.source))\(whereText):"
    return header + "\n" + (try renderBlock(node.body, indent + INDENT, markers))
}

private func renderWhen(_ node: AST.When, _ indent: String, _ markers: Markers) throws -> String {
    var entries: [String] = []
    for p in node.pattern {
        switch p.matcher {
        case let .match(arg): entries.append("\(p.field): \(try commaElement(arg, "record"))")
        case .bind: entries.append(p.field)
        }
    }
    let header = indent + "when \(try delimited(node.subject)) is { \(entries.joined(separator: ", ")) }:"
    var lines = [header, try renderBlock(node.body, indent + INDENT, markers)]
    if !node.els.isEmpty {
        lines.append(indent + "else:")
        lines.append(try renderBlock(node.els, indent + INDENT, markers))
    }
    return lines.joined(separator: "\n")
}

// The or-fail-with-handler at this statement's top, or nil.
private func statementOrfail(_ node: AST.Node) -> AST.OrFail? {
    if let o = node as? AST.OrFail, o.handler != nil { return o }
    if let a = node as? AST.Assign, let o = a.expr as? AST.OrFail, o.handler != nil { return o }
    if let g = node as? AST.Give, let o = g.expr as? AST.OrFail, o.handler != nil { return o }
    return nil
}

// A copy of `node` with its or-fail handler cleared, so the single-line render
// path produces the head line.
private func withoutHandler(_ node: AST.Node) -> AST.Node {
    func cleared(_ o: AST.OrFail) -> AST.OrFail { AST.OrFail(o.expr, o.tag, nil) }
    if let o = node as? AST.OrFail { return cleared(o) }
    if let a = node as? AST.Assign, let o = a.expr as? AST.OrFail {
        return AST.Assign(a.name, cleared(o), isLet: a.isLet, annotation: a.annotation)
    }
    if let g = node as? AST.Give, let o = g.expr as? AST.OrFail { return AST.Give(cleared(o)) }
    return node
}

public func renderStmt(_ node: AST.Node, _ indent: String, _ markers: Markers = Markers()) throws -> String {
    if let orfail = statementOrfail(node), let handler = orfail.handler {
        let head = try renderStmt(withoutHandler(node), indent, markers)
        let block = try renderBlock(handler, indent + INDENT, markers)
        return head + ":\n" + block
    }

    switch node {
    case let u as AST.Use: return indent + renderUse(u)
    case let f as AST.Foreign: return indent + renderForeign(f)
    case let r as AST.Rule: return indent + renderRule(r)
    case let n as AST.Note: return renderNote(n, indent)
    case let f as AST.FuncDef: return try renderFuncdef(f, indent, markers)
    case let a as AST.Assign: return indent + (try renderAssign(a))
    case let g as AST.Give: return indent + "give \(try renderExpr(g.expr))"
    case let s as AST.Show: return indent + "show \(try renderExpr(s.expr))"
    case let w as AST.Why: return indent + "why \(try renderExpr(w.expr))"
    case let w as AST.When: return try renderWhen(w, indent, markers)
    case let w as AST.WriteTo: return indent + (try renderWritetoInline(w))
    case let o as AST.OrFail: return indent + (try renderOrfail(o))
    case let f as AST.Fail: return indent + "fail \(try delimited(f.message)) as \(f.tag)"
    case let i as AST.If: return try renderIf(i, indent, markers)
    case let fe as AST.ForEach: return try renderForeachStmt(fe, indent, markers)
    default:
        if EXPR_STMT.contains(node.kind) { return indent + (try renderExpr(node)) }
        throw RenderError(message: "renderStmt: unhandled node type \(node.kind.rawValue)")
    }
}

// ================================================================ the generated marker

/// source line -> rule names, for every site an active rule reaches.
public struct Markers {
    var byLine: [Int: [String]] = [:]
    public init() {}
    public var isEmpty: Bool { byLine.isEmpty }
}

private func lineOfField(_ node: AST.Node) -> Int? {
    for f in node.fields where f.name == "line" {
        if case let .int(n) = f.value { return n }
    }
    return nil
}

// Every source line this node's subtree touches. Matches render.py's _line_span:
// recurses into direct node fields and one level into list/tuple elements that
// are themselves nodes (a tuple element, being no node, is not descended — a
// faithful quirk).
private func lineSpan(_ node: AST.Node, _ seen: inout Set<ObjectIdentifier>) -> Set<Int> {
    if !seen.insert(ObjectIdentifier(node)).inserted { return [] }
    var lines = Set<Int>()
    if let ln = lineOfField(node), ln != 0 { lines.insert(ln) }
    for f in node.fields {
        switch f.value {
        case let .node(v):
            lines.formUnion(lineSpan(v, &seen))
        case let .list(items), let .tuple(items):
            for case let .node(x) in items { lines.formUnion(lineSpan(x, &seen)) }
        default:
            break
        }
    }
    return lines
}

public func computeMarkers(_ rules: [AST.Rule]?, _ surface: Surface?, declaringFile: String? = nil) throws -> Markers {
    guard let rules, !rules.isEmpty else { return Markers() }
    guard let surface else {
        throw RenderError(message:
            "render(): rules given without surface -- markers need a " +
                "computed effect surface\n" +
                "  try: render(prog, rules=found, surface=analyse(src))")
    }
    var markers = Markers()
    for v in try check(rules, surface, declaringFile: declaringFile) {
        guard let effect = v.effect else { continue }
        markers.byLine[effect.site, default: []].append(v.rule.name)
    }
    return markers
}

private func markerLines(_ stmt: AST.Node, _ indent: String, _ markers: Markers) -> [String] {
    if markers.isEmpty { return [] }
    var seenNodes = Set<ObjectIdentifier>()
    let hit = lineSpan(stmt, &seenNodes).filter { markers.byLine[$0] != nil }.sorted()
    var names: [String] = []
    var seen = Set<CodePoints>()
    for ln in hit {
        for name in markers.byLine[ln]! where seen.insert(CodePoints(name)).inserted {
            names.append(name)
        }
    }
    return names.map { indent + "~ [\($0)] applies here" }
}

private func renderBlock(_ stmts: [AST.Node], _ indent: String, _ markers: Markers) throws -> String {
    var out: [String] = []
    for s in stmts {
        out.append(contentsOf: markerLines(s, indent, markers))
        out.append(try renderStmt(s, indent, markers))
    }
    return out.joined(separator: "\n")
}

// ================================================================ entry points

/// Canonical source text for a program. With `rules` and `surface`, governed
/// instruction sites carry a generated marker, computed here from `check`.
public func render(_ prog: [AST.Node], rules: [AST.Rule]? = nil, surface: Surface? = nil) throws -> String {
    let markers = try computeMarkers(rules, surface)
    let body = try renderBlock(prog, "", markers)
    return body.isEmpty ? "" : body + "\n"
}
